### PR TITLE
 Laboratory work #2, Matuschak Anastasia - 20FPL1 

### PR DESCRIPTION
**Pull request by almatuschak**

### CODE REVIEW

File: lab_2/main.py [hunk 1]
```diff
@@ -33,7 +60,15 @@ def get_language_features(language_profiles: dict) -> list or None:
         and sorts them in alphabetical order
     :param language_profiles: a dictionary of dictionaries - language profiles
     """
-    pass
+    if not isinstance(language_profiles, dict) or not language_profiles:
+        return None
+
+    features = []
+    for lang_profile in language_profiles.values():
+        for word in lang_profile.keys():
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:68]: what happens in this loop?

File: lab_2/main.py [hunk 2]
```diff
@@ -43,7 +78,20 @@ def get_text_vector(original_text: list, language_profiles: dict) -> list or Non
...
+    if not isinstance(original_text, list) or not isinstance(language_profiles, dict) or not language_profiles:
+        return None
+
+    features = get_language_features(language_profiles)
+    vector_list = []
+    for word in features:
+        if word in original_text:
+            for lang_profile in language_profiles.values():
+                if word in lang_profile.keys():
+                    vector_list.append(lang_profile[word])
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:90]: there is a way to get a key if it exists or 0 if it doesn't

File: lab_2/main.py [hunk 3]
```diff
@@ -47,19 +45,18 @@ def get_language_profiles(texts_corpus: list, language_labels: list) -> dict or
             return None
 
     language_profiles = {}
-    for label in range(len(language_labels)):
-        language_profiles[language_labels[label]] = get_freq_dict(texts_corpus[label])
-
+    for i, label in enumerate(language_labels):
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:48]: better to use names like index, token,etc.

File: lab_2/main.py (hunk 2 above)
Comment (ValeriyaKuznetsova) [reply lab_2/main.py:88]: google what arguments `get` needs
this is your task till Friday

File: lab_2/main.py [hunk 4]
```diff
@@ -101,7 +98,21 @@ def calculate_distance(unknown_text_vector: list, known_text_vector: list) -> fl
...
+        if not isinstance(i, (int, float)):
+            return None
+    for i in known_text_vector:
+        if not isinstance(i, (int, float)):
+            return None
+
+    distance = 0
+    for number, vector in enumerate(unknown_text_vector):
+        distance += (vector - known_text_vector[number])**2
+    return round(math.sqrt(distance), 5)
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:115]: what happens in this function?

File: lab_2/main.py [hunk 5]
```diff
@@ -112,7 +123,30 @@ def predict_language_score(unknown_text_vector: list, known_text_vectors: list,
...
+            return None
+
+    distances = {}
+    nearest_language = []
+    for i, vector in enumerate(known_text_vectors):
+        distances[language_labels[i]] = \
+            calculate_distance(unknown_text_vector, known_text_vectors[i])
+    for key, value in distances.items():
+        if value == min(distances.values()):
+            nearest_language = [key, value]
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:148]: what happens in this loop?

File: lab_2/main.py [hunk 6]
```diff
@@ -137,7 +184,40 @@ def predict_language_knn(unknown_text_vector: list, known_text_vectors: list,
...
+    knn_language = [language_labels[distances.index(distance)] for distance in knn_distances]
+
+    predict_language = {}
+    for language in knn_language:
+        if language not in predict_language:
+            predict_language[language] = 0
+        predict_language[language] += 1
+
+    predict_language = max(predict_language, key=predict_language.get)
+    return [predict_language, knn_distances[0]]
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:220]: what happens in this function?

File: lab_2/main.py (hunk 4 above)
Comment (almatuschak) [reply lab_2/main.py:115]: The distance is counted with euclid metric. At first, distance is equal to 0. For every vector in unknown_text_vector we add to distance the following: vector (from unknown_text_vector) - a vector from known_text_vector squared. Then the square root from this number is counted. We take only 5 digits of this number (with the round function).

File: lab_2/main.py (hunk 2 above)
Comment (almatuschak) [reply lab_2/main.py:88]: dict.get method gets a value from the dictionary by its key

Comment (almatuschak) [reply lab_2/main.py:88]: Changed to:
    for word in features:
        for lang_profile in language_profiles.values():
            if word in lang_profile:
                vector_list.append(lang_profile.get(word) if word in original_text else 0)

**Pull request closed without merging**